Container runtime env docs (#41764)
ray-project/ray: doc/source/serve/advanced-guides/multi-app-container.md

(serve-container-runtime-env-guide)=
# Run Multiple Applications in Different Containers

> This section goes over how you can manage dependencies per application in a more isolated manner by running your Serve applications in separate containers with different images.

Reviewer: ```suggestion
This section explains how to run multiple Serve applications on the same cluster in separate containers with different images.
```

:::note
This is an experimental feature and the API is subject to change. If you have additional feature requests or run into issues, please submit them on [Github](https://github.com/ray-project/ray/issues).
:::

## Install podman

> The `container` runtime environment feature uses [Podman](https://podman.io/) to start and run containers. Please follow these steps to install Podman in your environment. For instance, if you are using Kuberay, you need to use an image that has podman installed for the head and worker group specs.

Reviewer: ```suggestion
The `container` runtime environment feature uses [Podman](https://podman.io/) to start and run containers.
Please follow these steps to install Podman in the environment for all head and worker nodes.
```

> ::::{tab-set}

Reviewer 2: Non-blocker: I kinda feel this doesn't need to be a separate tabs. Just link directly to the Podman doc should cover most case. And can probably just add a note for Ubuntu 10.04 or older if we think that's important.

Author: Thanks for the suggestion, makes sense!

:::{tab-item} Ubuntu 20.10 and newer
```bash
sudo apt-get update
sudo apt-get install podman -y
```
:::

:::{tab-item} Ubuntu 20.04 and older
The podman package is only available in the official repositories for Ubuntu 20.10 and newer. To install podman in Ubuntu 20.04, you need to first add the software repository as a debian source.

```bash
sudo sh -c "echo 'deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04/ /' > /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list"
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 4D64390375060AA4
sudo apt-get update
sudo apt-get install podman -y
```
:::

:::{tab-item} Other
See [Podman Installation Instructions](https://podman.io/docs/installation).
:::

::::

> ## Run Serve application in a container

Reviewer: ```suggestion
## Run a Serve application in a container
```

> In this example, we are going to walk through deploying two applications in separate containers: a whisper model and a Resnet50 image classification model. 

Reviewer: ```suggestion
This example walks through deploying two applications in separate containers: a whisper model and a Resnet50 image classification model. 
```

> First, let's set up the docker images with the required dependencies. 

Reviewer: ```suggestion
First, install the required dependencies in the images.
```

:::note

> The Ray version and Python version in the container should match those of the host environment exactly. Note that for Python, the versions must match down to the patch number.

Reviewer: ```suggestion
The Ray version and Python version in the container *must* match those of the host environment exactly. Note that for Python, the versions must match down to the patch number.
```

Reviewer: I'd make this a `::: warning` so it pops (very important detail)

:::

In `whisper.Dockerfile`:
* We are running the Whisper model on GPU, so the base image for the Whisper Dockerfile is the latest Ray GPU image, `rayproject/ray:latest-py38-gpu`.
* We install the package `faster_whisper`, which is a dependency for the whisper model.
* We then download the source code for the whisper application into `whisper_example.py`.
Similarly, `resnet.Dockerfile` uses the appropriate base image, installs necessary dependencies, and downloads the resnet model source code.

> Save the following to files named `whisper.Dockerfile` and `resnet.Dockerfile`.

Reviewer: ```suggestion
`whisper.Dockerfile` does the following:
* Uses the latest Ray GPU image, `rayproject/ray:latest-py38-gpu`, so the Whisper model can run on GPUs.
* Installs the package `faster_whisper`, which is a dependency for the whisper model.
* Download the source code for the whisper application into `whisper_example.py`.

Similarly, `resnet.Dockerfile` uses the appropriate base image, installs necessary dependencies, and downloads the resnet model source code.
Save the following to files named `whisper.Dockerfile` and `resnet.Dockerfile`.
```

Reviewer: Might be more concise to annotate the dockerfiles with inline comments instead of describing them in prose

::::{tab-set}
:::{tab-item} whisper.Dockerfile
```dockerfile
FROM rayproject/ray:latest-py38-gpu

RUN pip install faster_whisper==0.10.0
RUN sudo apt-get update && sudo apt-get install curl -y

RUN curl -O https://raw.githubusercontent.com/ray-project/ray/master/doc/source/serve/doc_code/whisper_example.py
```
:::
:::{tab-item} resnet.Dockerfile
```dockerfile
FROM rayproject/ray:latest-py38-cpu

RUN torch==2.0.1 torchvision==0.15.2
RUN sudo apt-get update && sudo apt-get install curl -y

RUN curl -O https://raw.githubusercontent.com/ray-project/ray/master/doc/source/serve/doc_code/resnet50_example.py
```
:::
::::

Then, build the corresponding docker images and push it to your choice of docker registry.

::::{tab-set}
:::{tab-item} Whisper
```bash
export IMG1={your-image-repo-and-name}
docker build -t $IMG1 -f whisper.Dockerfile .
docker push $IMG1
```
:::
:::{tab-item} Resnet
```bash

> export IMG2={your-image-repo-and-name}

Reviewer 2: I feel for people who's not familiar with building and pushing docker images might not aware `your-image-repo-and-name` for `IMG1` and `IMG2` should be different. Maybe name them `whisper_image:latest` and `restnet_image:latest` or something alone those lines.

Author: Yup - I added a note about running locally, and I also changed the image names to `whisper_image:latest` and `resnet_image:latest`, adding a note that the user should change the names if they need to.

docker build -t $IMG2 -f resnet.Dockerfile .
docker push $IMG2
```
:::
::::

Finally, you can specify the docker image within which you want to run each application in the `container` field of an application's runtime environment specification. The `container` field has three fields:
- `image`: (Required) The image to run your application in.

> - `worker_path`: The absolute path to `default_worker.py` inside the container. This is necessary if the Ray installation directory is different than that of the host (where raylet is running). For instance, if you are running on bare metal, and using a standard Ray docker image as the base image for your application's, then the Ray installation directory on host will likely differ from that of the container. Note that if you are using a standard Ray docker image as the base image, then `default_worker.py` will always be found at `/home/ray/anaconda3/lib/{python-version}/site-packages/ray/_private/workers/default_worker.py`.

Reviewer: Does this mean if I use a standard Ray docker image, I don't need to set this field?

If so, let's call it out more clearly that this is an advanced option that most people shouldn't need to think about. Also, how can users debug if this is set incorrectly? What will the symptom be?

I'd consider omitting this field here and have an advanced section below describing it & how to debug/set it properly.

Author: If the user is running the Ray cluster in a container already, my guess is this is not necessary because they'll most likely use the same base image, or at least use the same method to install ray in the images, so that the installation path is the same.

> Also, how can users debug if this is set incorrectly? What will the symptom be?

I'm actually not certain, I'll check on this and add to the doc.

> I'd consider omitting this field here and have an advanced section below describing it & how to debug/set it properly.

I'm not 100% sure the "common" case doesn't require setting this. In a typical production setting, perhaps, since we probably expect them to be running the Ray cluster in a container already. However this isn't in the production guide, nor are we advertising that as the common case, so I want users who don't fit into that "typical production setting" case to be able to follow along as well. I can shorten the description here, then add an additional advanced section below like you suggested describing it in more detail and adding debugging instructions?

Reviewer: Shortening here and having a more detailed section w/ debugging sounds good to me

Author: Done!

- `run_options`: Additional options to pass to the `podman run` command used to start a Serve deployment replica in a container. See [podman run documentation](https://docs.podman.io/en/latest/markdown/podman-run.1.html) for a list of all options.

The following Serve config will run the `whisper` app with the image `your-image-1`, and the `resnet` app with the image `your-image-2`. Concretely, this means that all deployment replicas in the applications will start and run in containers with the respective images.

```yaml
applications:
  - name: whisper
    import_path: whisper_example:entrypoint
    route_prefix: /whisper
    runtime_env:
      container:
        image: {your-image-1}

>         worker_path: /home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/workers/default_worker.py

Reviewer: hopefully this is optional. it's very ugly, exposing a private API (the organization of internals of our python package)

Reviewer: why is this strictly necessary? Can't we do an `import ray` and find the path dynamically?

Reviewer: at a minimum, we should change this to only be the `/home/ray/anaconda3/lib/python3.8/site-packages/ray/` prefix so it isn't depending on the structure of the files in the `ray` package

Author: Agreed, this is pretty ugly. This is leftover from the initial implementation of this plugin, and I didn't have a chance to refine it. I'll focus on improving this UX in the upcoming release.

### Running the Ray cluster in a Docker container

If raylet is running inside a container, then that container will need the necessary permissions to start a new container. In order to do that, you need to start the container that runs raylet with the flag `--privileged`.

### Troubleshooting
* **Permission denied: '/tmp/ray/session_2023-11-28_15-27-22_167972_6026/ports_by_node.json.lock'**
  * If you're seeing this error, that likely means the user running inside the podman container is different from the host user that started the Ray cluster. The folder `/tmp/ray`, which is volume mounted into the podman container, is owned by the host user that started Ray, and the container is started with the `--userns=keep-id`, so there should only be permimssions issues if the user running inside the podman container is different. For instance, if the user on host is `root`, and you're using a container whose base image is a standard Ray image, then by default the container will be started with user `ray(1000)`, who won't be able to access the mounted `/tmp/ray` volume.

> * **ERRO[0000] 'overlay' is not supported over overlayfs: backing file system is unsupported for this graph driver**

Author: Yeah, overlayfs is a thing - it's used by containers to manage image layers.

>   * [This should only occur when you are running the Ray cluster inside a container] If you see this error when starting the replica actor, try volume mounting `/var/lib/containers` in the container that runs raylet. (i.e, add `-v /var/lib/containers:/var/lib/containers` to the command that starts the docker container)

Reviewer 3: ```suggestion
  * This error should only occur when you are running the Ray cluster inside a container. If you see this error when starting the replica actor, try volume mounting `/var/lib/containers` in the container that runs raylet. That is, add `-v /var/lib/containers:/var/lib/containers` to the command that starts the Docker container.
```

* **cannot clone: Operation not permitted; Error: cannot re-exec process**

>   * [This should only occur when you are running the Ray cluster inside a container] This implies that you don't have the permissions to use podman to start a container. You need to start the container (that runs raylet) with privileged permissions by adding `--privileged`.

Reviewer 3: ```suggestion
  * This error should only occur when you are running the Ray cluster inside a container. This error implies that you don't have the permissions to use Podman to start a container. You need to start the container that runs raylet, with privileged permissions by adding `--privileged`.
```
